### DISCUSSION TITLE
Auto-generated `tsconfig.json`

### DISCUSSION THREAD

**Discussion by florian-lefebvre** [withastro/roadmap#826] [Proposal]
### Body

# Summary

To improve DX for users, core and integrations authors, auto-generating a `tsconfig.json` inside `.astro` would allow to manage types for the user (through codegen) easily.

# Background & Motivation

Historically, the `src/env.d.ts` file has been really important in Astro to get Vite client and Content Collections types accessible in users' projects.

But as more and more features (both in core and integrations) will require codegen, it will add a new reference to `env.d.ts` for each one. I don't think that's necessary, that's an implementation detail. If users want their project to work correctly, they'll need those lines anyway.

So the idea of this proposal is to remove the need for this file (although keeping it as a convention for extending globals like window sounds pretty useful) by referencing those types with a little bit of *magic*. More details in the example below.

# Goals

A **concise, bulleted-list** outlining the intended goals of this RFC. 

- Remove the need for users to keep mandatory references inside `env.d.ts`
- Add a core utility to inject types through codegen
- Add an integration utility to inject types through codegen

The last point may require a separate rfc. FYI, we are experimenting with it in Astro Integration Kit, [see docs](https://astro-integration-kit.netlify.app/utilities/add-dts/).

# Example

## In other tools

I think nuxt is once again a pretty good example, [see their docs](https://nuxt.com/docs/guide/directory-structure/tsconfig). TLDR a basic user tsconfig looks like:

```json
{
  "extends": "./.nuxt/tsconfig.json"
}
```

This has a drawback though: it would mean loosing the ability for users to specify how strict they want TS to be.

## In Astro

Hopefully, there's a solution to this. Since TS 5.0, a tsconfig `extends` property can be an array. That's supported in astro thanks to https://github.com/withastro/astro/pull/8798.

So here is the idea: we can keep the extended tssconfig chosen by the user and also reference an auto-generated one:
```diff
// tsconfig.json
{
-	"extends": "astro/tsconfigs/base"
+	"extends": ["astro/tsconfigs/base", "./.astro/tsconfig.json"]
}
```
```diff
// src/env.d.ts
-/// <reference types="astro/client" />
-/// <reference types="../.astro/types.d.ts" />
```
```json
// .astro/tsconfig.json
{
	"compilerOptions": {
		"types": ["astro/client", "./types.d.ts"]
	}
}
```
I think it would be great if injecting some types using this new utility (from core or an integration) added a new `dts` file (for the sake of debugging), and this file was added to the `types` array of the generated tsconfig.

**Comment by Princesseuh** [comment]
You can't use `types` because you'll make type acquisition explicit, breaking the usage of global types like `@types/node`, you have to use `include`. 

However, the latter can be overriden by the root config, so that's something to watch out for. Nuxt "solves" this by including every single file, and prevent users from controlling `include` and `exclude` from `tsconfig.json` completely, which is a bit unfortunate, but I guess we could have a `typescript: {}` thing in our config where you can set it.

**Reply by florian-lefebvre** [reply]
Ah yeah I remember having this convo, I just...forgot about it. That's unfortunate to have to do this from an option in the astro config, but not the end of the world I guess. Technically a breaking change?

**Comment by Princesseuh** [comment]
![Screenshot_2024-02-02-16-36-38-044_com.discord-edit.jpg](https://github.com/withastro/roadmap/assets/3019731/3d83b384-0dc7-4a5f-b32e-41455d03455e)

Just reposting a comment on how I would suggest implementing this

**Comment by florian-lefebvre** [comment]
WIP PR at https://github.com/withastro/astro/pull/9952
